cicecore/shared/ice_distribution.F90

```diff
@@ -642,81 +647,40 @@ function create_distrb_cart(nprocs, workPerBlock) result(newDistrb)
+            localID = newDistrb%blockCnt(processor)
             newDistrb%blockLocation(globalID) = processor
             newDistrb%blockLocalID (globalID) = localID
-            newDistrb%blockCnt(processor) = newDistrb%blockCnt(processor) + 1
-            newDistrb%blockIndex(processor,localID) = globalID
          else  ! no work - eliminate block from distribution
             newDistrb%blockLocation(globalID) = 0
             newDistrb%blockLocalID (globalID) = 0
          endif
       end do
       end do
 
-      ! if this is the local processor, set number of local blocks
-      if (my_task == processor - 1) then
-         newDistrb%numLocalBlocks = localID
-      endif
-
    end do
    end do
 
-!----------------------------------------------------------------------
-!
-!  now store the local info
-!
-!----------------------------------------------------------------------
+   newDistrb%numLocalBlocks = newDistrb%blockCnt(my_task+1)
 
-   if (newDistrb%numLocalBlocks > 0) then
-      allocate (newDistrb%blockGlobalID(newDistrb%numLocalBlocks), &
-                stat=istat)
-      if (istat > 0) then
-         call abort_ice( &
-            'create_distrb_cart: error allocating blockGlobalID')
-         return
+   ! set local blockGlobalID array
+   allocate(newDistrb%blockGlobalID(newDistrb%numLocalBlocks), stat=istat)
+   if (ice_memusage_allocErr(istat,subname//'alloc numLocalBlocks')) return
+   do n = 1,nblocks_tot
+      if (my_task+1 == newDistrb%blockLocation(n)) then
+         localID = newDistrb%blockLocalID(n)
+         newDistrb%blockGlobalID (localID) = n
       endif
+   enddo
 
-      do j=1,nprocsY
-      do i=1,nprocsX
-         processor = (j-1)*nprocsX + i
-
-         if (processor == my_task + 1) then
-            is = (i-1)*numBlocksXPerProc + 1   ! starting block in i
-            ie =  i   *numBlocksXPerProc       ! ending   block in i
-            if (ie > nblocks_x) ie = nblocks_x
-            js = (j-1)*numBlocksYPerProc + 1   ! starting block in j
-            je =  j   *numBlocksYPerProc       ! ending   block in j
-            if (je > nblocks_y) je = nblocks_y
-
-            localID        = 0  ! initialize counter for local index
-            do jblock = js,je
-            do iblock = is,ie
-               globalID = (jblock - 1)*nblocks_x + iblock
-               if (workPerBlock(globalID) /= 0) then
-                  localID = localID + 1
-                  newDistrb%blockGlobalID (localID) = globalID
-               endif
-            end do
-            end do
-
-         endif
-
-      end do
-      end do
-
-   else
-      allocate (newDistrb%blockGlobalID(newDistrb%numLocalBlocks), &
-                stat=istat)
-      if (istat > 0) then
-         call abort_ice( &
-            'create_distrb_cart: error allocating blockGlobalID')
-         return
+   ! set/check max_blocks
+   if (lmax_blocks_calc) then
+      if (max_blocks < 0) then
+         max_blocks = newDistrb%numLocalBlocks
```

Comment:
max_blocks is used throughout the code as the number of blocks to allocate on each task.  It used to be the same value on all tasks, set by the user.  Now it's computed internally and set uniquely for each task.  Setting max_blocks in line 683 is doing exactly that.  Because the rake decomp calls the cartesian decomp, I needed to add an option where I could turn off the setting of max_blocks in cart because I calculate it in rake.  But if cart is called directly, then max_blocks is set.  Every decomp option sets the max_blocks variable.

Reply:
I am confused here. I thought nblocks was meant to be the number of blocks on each task and this could be different. The parameter max_blocks should be the maximum number of blocks across all tasks, no?

Reply:
@dabail10, you are correct.  In general, nblocks is the number of active blocks on each task.  max_blocks is the number of blocks that are allocated on each task.  Historically, max_blocks was used to allocate memory, was the same value on all tasks, and was set by a CPP at build time.  While nblocks was used to loop over only active blocks at run time.  The code still uses those two variables as they've always been defined.

But in the last few years, we moved to dynamic allocation of memory (moving max_blocks to namelist) and with the current PR, we are able to compute max_blocks internally on each task BEFORE we need to use it to allocate memory.  So in that case, nblocks and max_blocks can overlap in function.

However, we still support user defined max_blocks (although we probably don't need to) which means we still want to differentiate max_blocks and nblocks.

Maybe the next step is to ignore the max_blocks namelist setting and always set it internally.  Once we do that, we could, in theory, unify nblocks and max_blocks in the code to a single variable.  But we're not quite there yet, and I think we could debate whether all that refactoring would be worth the effort.  max_blocks and nblocks still play different roles (memory allocation vs active block count), and we're all pretty comfortable with that scheme.

Reply:
Having a "max_blocks" in the namelist and a "max_blocks" in the code with different meanings is confusing. I see that it may not be worth it to unify max_blocks and nblocks (there are ~1000 uses of max_blocks), so I guess we just run with it as is.

This comment in `ice_domain_size` is now wrong:

        max_blocks  , & ! max number of blocks per processor

and maybe could be _number of blocks memory is allocated for_ or similar?